### PR TITLE
feat/AUT-1659/enable plugins from presets list in Test Preview

### PR DESCRIPTION
**Pull request by bartlomiejmarszal**
Related to: https://oat-sa.atlassian.net/browse/AUT-1659
**BE part:**
TestPreviewer::init will now include `presetList` variable that will map all tools available and their map to QTI categories
**FE part:** 
on actions testContext.properties updated using preset categories to enable plugins in Test Preview

**How to test:**
- create test
- add item
- enable some plugins for item (Test-taker tools)
- save and go to test preview
- plugins should be enabled

image
image

### CODE REVIEW

File: models/TestCategoryPresetMap.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    private $testCategoryPresetProvider;
+
+    public function __construct(TestCategoryPresetProvider $testCategoryPresetProvider)
+    {
+        $this->testCategoryPresetProvider = $testCategoryPresetProvider;
+    }
+
+    public function getMap(): array
+    {
+        $presetList = $this->testCategoryPresetProvider
```
Comment (andreluizmachado) [inline models/TestCategoryPresetMap.php:44]: What happens if the find fails here?  Do we need to catch the exception?

File: test/unit/models/TestCategoryPresetMapTest.php [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    public function setUp(): void
+    {
+        $this->testCategoryPresetProviderMock = $this->createMock(TestCategoryPresetProvider::class);
+        $this->testCategoryPresetMock = $this->createMock(TestCategoryPreset::class);
+        $this->subject = new TestCategoryPresetMap(
+            $this->testCategoryPresetProviderMock
+        );
+    }
+
+    public function testGetMap(): void
```
Comment (andreluizmachado) [inline test/unit/models/TestCategoryPresetMapTest.php:52]: is it possible to add a test case for an empty result?

File: views/js/previewer/proxy/test.js [hunk 3]
```diff
@@ -62,17 +62,39 @@ define([
...
+     * @param {Number} position - item position
+     * @param {Object} testContext
+     * @param {Array} presetMap
      */
-    function findIds(testMap, position) {
-        const item = mapHelper.getJump(testMap, position);
+    function updateTestContextWithItem(testMap, position, testContext, presetMap) {
+        const jump = mapHelper.getJump(testMap, position);
+        const item = mapHelper.getItemAt(testMap, position);
         if (item) {
```
Comment (andreluizmachado) [inline views/js/previewer/proxy/test.js:74]: you can use [early returns ](https://dev.to/yvonnickfrin/devtips-use-early-returns-to-avoid-nested-conditions-4i51)here.

File: views/js/previewer/proxy/test.js [hunk 4]
```diff
@@ -62,17 +62,39 @@ define([
...
+            testContext.itemSessionState = itemSessionStates.initial;
+            testContext.options = createContextOptions(item, presetMap);
         }
-        return {};
+    }
+    /**
+     * Convert preset categories to context.options
+     * @param {Object} item
+     * @param {Array} presetMap
+     * @returns {Object} properties
```
Comment (andreluizmachado) [inline views/js/previewer/proxy/test.js:86]: wouldn't be better: @returns {Object} options?

File: models/TestCategoryPresetMap.php (hunk 1 above)
Comment (bartlomiejmarszal) [reply models/TestCategoryPresetMap.php:44]: It will throw exception, we will catch it on controller level

File: views/js/previewer/proxy/test.js [hunk 5]
```diff
@@ -159,17 +183,17 @@ define([
...
                 })
-                .then(data => {
-                    data.itemData = data.content;
-                    data.itemIdentifier = data.content.data.identifier;
-                    data.itemState = {};
-                    this.itemStore[itemIdentifier] = data;
-                    return data;
-                })
-                .then(itemData => itemDataHandlers(itemData));
+                    .then(data => {
```
Comment (andreluizmachado) [inline views/js/previewer/proxy/test.js:189]: is this indentation correct?

File: test/unit/models/TestCategoryPresetMapTest.php (hunk 2 above)
Comment (bartlomiejmarszal) [reply test/unit/models/TestCategoryPresetMapTest.php:52]: We could but a behavior will be not different from current test

Comment (andreluizmachado) [reply test/unit/models/TestCategoryPresetMapTest.php:52]: The idea here would be to cover both cases. This one is not covered.

File: models/TestCategoryPresetMap.php (hunk 1 above)
Comment (andreluizmachado) [reply models/TestCategoryPresetMap.php:44]: I see. In this case, Do we really need to stop the entire flow because of it? Is it possible to catch it here by logging the error and doing a fallback so the application doesn't crash?

Comment (bartlomiejmarszal) [reply models/TestCategoryPresetMap.php:44]: I think failure in this case would signal critical issue that will impact instance. I could send empty array but FE will not be able to handle with some of their plugins. @bziondik?

File: test/unit/models/TestCategoryPresetMapTest.php (hunk 2 above)
Comment (bartlomiejmarszal) [reply test/unit/models/TestCategoryPresetMapTest.php:52]: I do not see too much value of that test but I did add it.

File: models/TestCategoryPresetMap.php (hunk 1 above)
Comment (bziondik) [reply models/TestCategoryPresetMap.php:44]: I think it is critical issue

File: models/ServiceProvider/QtiTestPreviewerServiceProvider.php [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2021 (original work) Open Assessment Technologies SA;
```
Comment (hectoras) [inline models/ServiceProvider/QtiTestPreviewerServiceProvider.php:18]: nitpick: `year++` :)

```suggestion
 * Copyright (c) 2022 (original work) Open Assessment Technologies SA;
```

File: models/TestCategoryPresetMap.php [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2021 (original work) Open Assessment Technologies SA;
```
Comment (hectoras) [inline models/TestCategoryPresetMap.php:18]: nitpick: `year++` :)

```suggestion
 * Copyright (c) 2022 (original work) Open Assessment Technologies SA;
```

File: test/unit/models/TestCategoryPresetMapTest.php [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2021 (original work) Open Assessment Technologies SA;
```
Comment (hectoras) [inline test/unit/models/TestCategoryPresetMapTest.php:18]: nitpick: `year++` :)

```suggestion
 * Copyright (c) 2022 (original work) Open Assessment Technologies SA;
```

File: views/js/previewer/proxy/test.js (hunk 3 above)
Comment (bziondik) [reply views/js/previewer/proxy/test.js]: fixed

File: views/js/previewer/proxy/test.js (hunk 5 above)
Comment (bziondik) [reply views/js/previewer/proxy/test.js:190]: yes

**Review by Silvia-loza — APPROVED** [review]
FE Review and local  test of feature
- [ ] New code is covered by tests (if applicable)
- [X] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [X] New code is respecting code style rules
- [X] New code is respecting best practices
- [X]  Pull request's target is not` master`
- [X] Feature is working correctly on my local machine (if applicable)
- [X] Acceptance criteria are respected
- [X] Pull request title and description are meaningful

**Review by pnal — APPROVED** [review]
Good job!
- [X] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [X] New code is respecting code style rules
- [X] New code is respecting best practices
- [X]  Pull request's target is not` master`
- [X] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [X] Pull request title and description are meaningful

**Comment by AntonNeviarouski** [comment]
QA Verified
